### PR TITLE
[W5.7][W16-4] Bryan Lew

### PR DESCRIPTION
**Pull request by blewjy**
Added 'findtag' command which find persons by tag name.
Added relevant tests.
Updated user guide.

### CODE REVIEW

**Comment by ChesterSng** [comment]
Very useful addition to the use of tags

**Comment by clawyq** [comment]
Maybe the functionality of your current find tag can be generalised to a broader 'filter-like' method, allowing you to not just filter for a specific keyword in the tag field, but maybe also in the name, email and address fields as well.

Code is easy to understand and cleanly written.

**Comment by 99percentile** [comment]
Might be nice if you include a command to list all tags in the address book.

But this is useful as well!

File: src/seedu/addressbook/commands/HelpCommand.java [hunk 1]
```diff
@@ -15,14 +15,14 @@
     public CommandResult execute() {
         return new CommandResult(
                 AddCommand.MESSAGE_USAGE
-                + "\n" + DeleteCommand.MESSAGE_USAGE
```
Comment (99percentile) [inline src/seedu/addressbook/commands/HelpCommand.java:18]: Why did you do this?

File: src/seedu/addressbook/commands/FindTagCommand.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        return new CommandResult(getMessageForPersonListShownSummary(personsFound), personsFound);
+    }
+
+    /**
+     * Retrieves all persons in the address book whose tags are contained in the set of keywords.
+     *
+     * @param keywords for searching
+     * @return list of persons found
+     */
+    private List<ReadOnlyPerson> getPersonsWithNameContainingAnyKeyword(Set<String> keywords) {
```
Comment (bqnguyen94) [inline src/seedu/addressbook/commands/FindTagCommand.java:45]: I thought the method looks for `tag`?

File: src/seedu/addressbook/commands/ViewAllCommand.java [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
     public static final String COMMAND_WORD = "viewall";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Views the non-private details of the person "
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Views all (including private) details of the person "
```
Comment (bqnguyen94) [inline src/seedu/addressbook/commands/ViewAllCommand.java:15]: Is this related?

File: test/java/seedu/addressbook/common/UtilsTest.java [hunk 4]
```diff
@@ -11,6 +11,36 @@
...
+        // multiple objects with null
+        assertHasNull(null, 1, Integer.valueOf(1));
+        assertHasNull(null, null);
+        assertHasNull(null, "a", "b", null);
+
+    }
+
+    private void assertHasNull(Object... objects) { assertTrue(Utils.isAnyNull(objects)); }
+
+    private void assertNoNull(Object... objects) { assertFalse(Utils.isAnyNull(objects)); }
```
Comment (bqnguyen94) [inline test/java/seedu/addressbook/common/UtilsTest.java:42]: This null test is not relevant to this PR, so please remove it :)

File: test/java/seedu/addressbook/parser/ParserTest.java [hunk 5]
```diff
@@ -197,6 +187,46 @@ public void parse_findCommandDuplicateKeys_parsedCorrectly() {
...
+        final String[] keywords = { "key1", "key2", "key3" };
+        final Set<String> keySet = new HashSet<>(Arrays.asList(keywords));
+
+        // duplicate every keyword
+        final String input = "findtag " + String.join(" ", keySet) + " " + String.join(" ", keySet);
+        final FindTagCommand result =
+                parseAndAssertCommandType(input, FindTagCommand.class);
+        assertEquals(keySet, result.getKeywords());
+    }
+
```
Comment (bqnguyen94) [inline test/java/seedu/addressbook/parser/ParserTest.java:229]: For this command you can test executing the command as well, but for now I think this is satisfactory.

**Review by bqnguyen94 — CHANGES_REQUESTED** [review]
Good job 👍 some changes needed.

**Pull request closed without merging**